pkg/services/ngalert/schedule/schedule.go

```diff
@@ -315,11 +324,22 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 	evalDuration := sch.metrics.EvalDuration.WithLabelValues(orgID)
 	evalTotalFailures := sch.metrics.EvalFailures.WithLabelValues(orgID)
 
-	clearState := func() {
-		states := sch.stateManager.ResetStateByRuleUID(grafanaCtx, key)
-		expiredAlerts := FromAlertsStateToStoppedAlert(states, sch.appURL, sch.clock)
-		if len(expiredAlerts.PostableAlerts) > 0 {
-			sch.alertsSender.Send(key, expiredAlerts)
+	clearState := func(ctx context.Context, err error) {
+		// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted/paused
+		isRuleDeleted := errors.Is(err, errRuleDeleted)
+		isRulePaused := errors.Is(err, errRulePaused)
+		if isRuleDeleted || isRulePaused {
+			var states []*state.State
+			rule := sch.schedulableAlertRules.get(key)
```

Comment:
There is a test `TestProcessTicks` where `rule` is `nil` and causes a panic

Reply:
I think the issue here is that the test deletes the alert rule and then calls `processTick`. The scheduler needs to be able to stop rule evaluation when a rule has been deleted and is no longer in the registry.

Reply:
I'll have a look at that, thanks

Reply:
FIxed, thanks again

Reply:
I would probably move Line 344 to inside the if statement on Line 347.

Reply:
You are totally right, should have done it when I removed the rule argument from `ResetStateByDeletedRuleUID`